src/client/gui/lib/vm_details/terminal_tabs.dart

```diff
@@ -81,90 +124,65 @@ class Tab extends StatelessWidget {
   }
 }
 
-class TerminalTabs extends StatefulWidget {
+class TerminalTabs extends ConsumerWidget {
   final String name;
 
   const TerminalTabs(this.name, {super.key});
 
   @override
-  State<TerminalTabs> createState() => _TerminalTabsState();
-}
-
-class _TerminalTabsState extends State<TerminalTabs> {
-  var _currentIndex = 0;
-
-  get currentIndex => _currentIndex;
-
-  set currentIndex(value) {
-    final (_, key) = shells[value];
-    FocusManager.instance.primaryFocus?.unfocus();
-    key.currentState?.focusNode.requestFocus();
-    _currentIndex = value;
-  }
-
-  final shells = [(1, GlobalKey<VmTerminalState>())];
+  Widget build(BuildContext context, WidgetRef ref) {
+    final provider = shellIdsProvider(name);
+    final notifier = provider.notifier;
+    final (:ids, :currentIndex) = ref.watch(provider);
 
-  @override
-  Widget build(BuildContext context) {
-    final tabs = shells.indexed.map((e) {
-      final (orderIndex, (shellId, _)) = e;
-      return ReorderableDragStartListener(
+    final tabsAndShells = ids.mapIndexed((index, shellId) {
+      final tab = ReorderableDragStartListener(
         key: ValueKey(shellId),
-        index: orderIndex,
+        index: index,
         child: Tab(
           title: 'Shell $shellId',
-          selected: orderIndex == currentIndex,
-          onTap: () => setState(() => currentIndex = orderIndex),
-          onClose: () {
-            if (shells.length == 1) return;
-            setState(() {
-              shells.removeAt(orderIndex);
-              if (orderIndex < currentIndex) currentIndex -= 1;
-              currentIndex = currentIndex.clamp(0, shells.length - 1);
-            });
-          },
+          selected: index == currentIndex,
+          onTap: () => ref.read(notifier).setCurrent(index),
+          onClose: () => ref.read(notifier).remove(index),
```

Comment:
Indeed, that's an issue. The idea is that, as you saw, if you close your last shell, it automatically opens a new one with ID 1. If you try to close the last one and it has ID 1, it is removed, but within the same operation, a new one with ID one is added. But because they happened basically at the same time, the framework does not differentiate between them, so it behaves as if shell with ID 1 never disappeared. I can fix it so that when you close your last shell and it has ID 1, it respawns a totally new Shell with ID instead of keeping the old one. But this bring into question how exactly should the terminal tabs behave. I've also took a note of this and will discuss it with Design. As of now, there is no design for what should be shown when there are 0 shells, that's why it automatically opens a new one if the last one is closed.